### ISSUE TITLE
Provide behavior options for time::Interval

### ISSUE THREAD

**Issue by Lindenk** [tokio-rs/tokio#3574]
**Is your feature request related to a problem? Please describe.**
Yes. The current implementation of `time::Interval` attempts to smooth missed ticks by "catching up" when it has the chance. This is not always wanted behavior. This would also be a solution to #2301

**Describe the solution you'd like**
There are three possible behaviors an interval can have when missing ticks:

***Catch Up***

The interval will schedule ticks immediately if it falls behind. This "smooths" the tick rate by ticking `time_passed / interval_duration` times even when some ticks are delayed as long as `total_work_time_required < time_passed`. If one tick is significantly delayed (let's say by `x` ticks), the interval will immediately trigger `x-1` ticks afterwards before returning to normal.

```
      1            2            3            4            5            6
|------------|------------|------------|------------|------------|------------|
| work |-----|            work            |  work  | work | work | work |-----|
```

***Delay***

The interval will schedule ticks at `max(prev_tick + duration, now)`. If a tick falls behind, the next one is scheduled immediately but future ticks are scheduled as normal. This effectively *delays* interval ticks by however much the overloaded tick fell behind. This can cause the interval to no longer be triggered at a multiple of it's duration but avoids many ticks being triggered at once after an overloaded tick.

```
      1                     2                   3            4            5            6
|------------|----------------------------|------------|------------|------------|------------|
| work |-----|            work            |  work  |---| work |-----| work |-----| work |-----|
```

***Skip***

The interval will always schedule ticks at the start of the next `start + x*duration`. This will result in *skipping* ticks if one takes too long but keeps in sync with predictable tick times without triggering a bunch of ticks at once to catch up.

```
      1            2            -            3            4            5
|------------|------------|------------|------------|------------|------------|
| work |-----|            work            |  work  || work |-----| work |-----|
```

**New implementation**

Ideally an `Interval` trait could provide an interface to `IntervalCatchup`, `IntervalDelay`, and `IntervalSkip` structs with `interval_at` returning `IntervalCatchup` by default, and a seperate `interval_at_with_behavior` function to select one of the others. Unfortunately async traits aren't really a thing at the moment.

Another solution could be to provide the above three structs and alias `Interval` to `IntervalCatchup` for the time being to preserve backwards comparability.

Or a better idea for the implementation, I'm not picky

**Describe alternatives you've considered**
Create my own implementations of `Interval` instead of using `tokio`'s

**Comment by Darksonn** [comment]
One option would be an `set_catchup` method that sets the strategy used by the `Interval` instead of adding three new types.

**Comment by malaire** [comment]
Your picture of **skip** doesn't match the description as 3rd work is not scheduled at start of next `start + x*duration`. I'd expect behavior to be:
```
     1            2            -            3            4            5
|------------|------------|------------|------------|------------|------------|
| work |-----|            work            |---------| work |-----| work |-----|
```

**Comment by Lindenk** [comment]
@malaire Well, the next tick is scheduled once a `Ready` is returned. When tick 2 triggers, it schedules the next tick one duration forward, the end of 2/beginning of -. The next time the interval is polled, it's way past due and would trigger immediately, scheduling the next tick at the start of 4. 

If it didn't trigger the past due tick immediately, even only slightly late polls would just reschedule their tick instead of returning `Ready`. I suppose it could timeout after more than one duration has passed and reschedule it's  tick when polled instead of becoming `Ready`. This could lead to interval ticks never becoming ready if they aren't polled often enough though

**Comment by malaire** [comment]
I don't know how hard that would be to implement, but that behavior described in my previous post would be most useful one for me.

**Comment by Lindenk** [comment]
Hmm, I suppose the problem could be broken down a bit further. What about having two configuration options on `Interval` (with better/more concise wording): `maximum_catchup_ticks` and `minimum_tick_length`. 

`maximum_catchup_ticks` would only trigger ticks early to catch up until it ticks the given number of times, then skips the rest. If it's set to `0`, there wouldn't be any catchup ticks and it would work the way @malaire describes, or if it's set to `1`, it would work the way it shows in my example. Defaults to no maximum to have the current behavior.

`minimum_tick_length` would force each tick to last for at least this duration (or scaling factor of the interval duration). If set to 0, it would have the behavior of *Catch Up*; ticks can be arbitrarily short when catching up. If set to 1 tick length, it would have the behavior of *Delay*; ticks will always be at least one full duration and never try to catch up. If set to something like 0.5 tick length, ticks will attempt to catch up, but only so fast; no tick will trigger faster than 0.5x the tick duration.

With these options you could do things like:

`maximum_catchup_ticks`: 2
`minimum_tick_length`: 0.8 * tick_duration

```
        1              2              3              4              5              -              6       
|--------------|--------------|--------------|--------------|--------------|--------------|--------------|
| work |-------|             work              |  work  |--| work |----| work |-----------| work |-------|
                         (.8 * of an interval) ^           ^  
                                               (skipped because catchup ticks = 2) ^
```

**Comment by malaire** [comment]
I don't understand that last example: Aren't there 3 catchup ticks, not 2?

**Comment by Lindenk** [comment]
@Darksonn Agreed.

@malaire Yep, you're right. Let me edit that

**Comment by sb64** [comment]
I can see two potential options for allowing users to specify the catchup method used by `Interval`:

- First, do what @Lindenk initially suggested and provide an `Interval` trait with three separate structs that implement the different suggested catchup behaviors. The only real pro of this that I can think of is that it might be slightly more efficient, as it would avoid branching on every tick. However, this seems to me to be the worse option because it would bloat the `time` module with more symbols, which would make it harder for users to understand. However, if we provided a default behavior, most users wouldn't face much additional complexity. Additionally, if the distinction was well-documented, users who needed to choose a specific behavior wouldn't experience too much frustration. However, as @Lindenk mentioned, `async` `trait`s are not a thing, so that would require a workaround.
- Secondly, do what @Darksonn suggested and provide a `set_behavior` and/or `interval_with_behavior` function(s) to allow the user to specify the catchup behavior they want the `Interval` to use. The behavior could be stored in an `enum`, and even provide more complex configuration options as @Lindenk suggested. Additionally, this solution would allow users to dynamically change the behavior that the `Interval` implements. However, this might cause more messy code because `tick` would need to `match` on the behavior chosen. Additionally, this would necessitate a branch for every tick. I'm not sure how performance-sensitive Tokio is intended to be, but that could theoretically cause a slight loss in performance.

I think the second option is the better one, but before I go ahead and implement this, I wanted to check with @Darksonn. As a maintainer, which do you think would be the better option?

**Comment by sb64** [comment]
I can see two potential options for allowing users to specify the catchup method used by `Interval`:

-First, do what @Lindenk initially suggested and provide an `Interval` trait with three separate structs that implement the different suggested catchup behaviors. The only real pro of this that I can think of is that it might be slightly more efficient, as it would avoid branching on every tick. However, this seems to me to be the worse option because it would bloat the `time` module with more symbols, which would make it harder for users to understand. However, if we provided a default behavior, most users wouldn't face much additional complexity. Additionally, if the distinction was well-documented, users who needed to choose a specific behavior wouldn't experience too much frustration. However, as @Lindenk mentioned, `async` `trait`s are not a thing, so that would require a workaround.
-Secondly, do what @Darksonn suggested and provide a `set_behavior` and/or `interval_with_behavior` function(s) to allow the user to specify the catchup behavior they want the `Interval` to use. The behavior could be stored in an `enum`, and even provide more complex configuration options as @Lindenk suggested. Additionally, this solution would allow users to dynamically change the behavior that the `Interval` implements. However, this might cause more messy code because `tick` would need to match on the behavior chosen. Additionally, this would necessitate a branch for every tick. I'm not sure how performance-sensitive Tokio is intended to be, but that could theoretically cause a slight loss in performance.

I think the second option is the better one, but before I go ahead and implement this, I wanted to check with @Darksonn. As a maintainer, which do you think is the better option?

**Comment by Darksonn** [comment]
I will note that the first option doesn't really require there to be a trait. You can have multiple `Interval` structs without there being a trait for it.

That said, I prefer option two.

**Comment by sb64** [comment]
That's true, but I figured that having a trait would make it easier for users to understand the role of the different structs.
However, that's a moot point, and I'll go ahead and start with option two.
